jdbc session handling in JSON format

davidblasby opened the pull request:
WIP - do not merge

This sets up JDBC Sessions so they read/write in JSON format (jsonb in postgresql).

1. make sure you have spring JDBC
     * sessions setup (cf application.yaml) 
 
 2. make sure to change the SPRING_SESSION_ATTRIBUTES column "ATTRIBUTE_BYTES"
     * from type "bytea" to "jsonb" 
    * see schema-postgresql-jdbc.sql 
 3. this will configure: 
     a. "springSessionConversionService" bean to configure the conversion 
     b. modify the load/save SQL to be postgresql jsonb aware 
    c. add trivial JsonDeserializer/JsonSerializer to do the conversion


Typical JSON stored (in PG DB table spring_session_attributes) after logging in a "admin/admin":

```
{
  "@class": "org.springframework.security.core.context.SecurityContextImpl",
  "authentication": {
    "@class": "org.springframework.security.authentication.UsernamePasswordAuthenticationToken",
    "details": {
      "@class": "org.springframework.security.web.authentication.WebAuthenticationDetails",
      "sessionId": "10c7a98d-8640-43cb-be8f-88cccb314de1",
      "remoteAddress": "0:0:0:0:0:0:0:1"
    },
    "principal": {
      "@class": "org.springframework.security.core.userdetails.User",
      "enabled": true,
      "username": "admin",
      "authorities": [
        "java.util.Collections$UnmodifiableSet",
        [
          {
            "@class": "org.springframework.security.core.authority.SimpleGrantedAuthority",
            "authority": "ROLE_Administrator"
          }
        ]
      ],
      "accountNonLocked": true,
      "accountNonExpired": true,
      "credentialsNonExpired": true
    },
    "authorities": [
      "java.util.Collections$UnmodifiableRandomAccessList",
      [
        {
          "@class": "org.springframework.security.core.authority.SimpleGrantedAuthority",
          "authority": "ROLE_Administrator"
        }
      ]
    ],
    "authenticated": true
  }
}
``` 

principal.username: "admin"
authorities[0].authority: ROLE_Administrator

david-blasby committed:
jdbc session

davidblasby commented on config/application.yml:
```diff
@@ -58,13 +58,18 @@ spring:
     name: GeoNetwork
   datasource:
-    url: jdbc:postgresql://localhost:5432/geonetwork
+    url: jdbc:postgresql://localhost:5432/gn
```
ignore this - this changes the database name from "geonetwork" to "gn"

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/geonetwork/geonetwork?pullRequest=75) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/geonetwork/geonetwork?pullRequest=75) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/geonetwork/geonetwork?pullRequest=75) it.</sub>

davidblasby commented:
@fxprunayre @josegar74 - ok, I've moved things around and put in an implementation for "/collections/{collectionId}".

1. the ISO samples to your GN
2. add the 2 records, attached
3. create a sub-portal and attach it to the "sub-portal" record (`serviceRecord`)
4. in admin->settings -> system/CSW/capabilityRecordUuid and set to the "portal" record
    * there isn't a way to attach a service record to the "main" portal in GN UI
5.  go to http://localhost:7979/ogcapi-records/collections/subportal?f=html (subportal=name you used for #3)
6. `select * from source` in the DB to get the id for the main portal.

I modified the IndexRecord so that geometries isn't a string, but a json object - this is causing some test case failures.  Most likely because I didn't update the XSLT.  This is something to look at in 2025.

[main (1).txt](https://github.com/user-attachments/files/18190448/main.1.txt)
[sub.txt](https://github.com/user-attachments/files/18190449/sub.txt)
